### DISCUSSION TITLE
Event hook on redirect request not being called twice

### DISCUSSION THREAD

**Discussion by kummerer94** [encode/httpx#1804] [Potential Issue] [answered]
Hi!

Thanks for the brillant library! I love the flexibility it offers when compared to `aiohttp`. It was much easier setting up something like a requests `HTTPAdapter` with this library than with `aiohttp`.

But I think that there could be some issue with the event hook or at least it would be great to clarify this in the documentation. Let's look at this example:

```python
import asyncio

import httpx


class AsyncTokenTransport(httpx.AsyncHTTPTransport):
    async def handle_async_request(self, method, url, headers, stream, extensions):
        print(f"Intercepted {url[1].decode()}")
        return await super().handle_async_request(
            method, url, headers, stream, extensions
        )


async def test_hook(request: httpx.Request) -> None:
    print(f"Called for request to {request.url}")


async def main():
    async with httpx.AsyncClient(
        transport=AsyncTokenTransport(), event_hooks={"request": [test_hook]}
    ) as client:
        # URL is the one of my last school which was handy because it has a redirect configured
        r = await client.get("http://cogym.de")
        print(r.status_code)


loop = asyncio.get_event_loop()
loop.run_until_complete(main())
```

The output is
```
Called for request to http://cogym.de
Intercepted cogym.de
Intercepted www.carl-orff-gym.de
200
```

This means, while the transport will be called for both requests, the hook will only be called once.

I have a use case where I need to attach an authentication token in the header for every request that needs to be unique for every request. Unfortunately, if I call an API route that redirects me (and attach the token via the event hook), `httpx` will automatically follow the redirect but reuse the previous headers. Therefore, I use the same authentication token twice, which leads to an auth error.

Could you clarify if this is the intended behaviour? Do you want the event hook to only be called once if we follow the redirect and make another request? That definitely was new to me and the requests `HTTPAdapter` that I used previously would have been called twice here.

If this is the intended behaviour, I think it would be great to clarify this in the docs. I am happy to contribute a PR.

**PS:**

As my example shows, you can do the thing that I want to do with a custom Transport. But this is not as easy or convenient as the event hook. In the event hook, I can work with the nice top-level `Request` instance (including the `URL` instance).

**Comment by lovelydinosaur** [comment, marked as answer]
Hola!

> Could you clarify if this is the intended behaviour? Do you want the event hook to only be called once if we follow the redirect and make another request? That definitely was new to me and the requests HTTPAdapter that I used previously would have been called twice here.

So... we've got a choice here, which I think is either...

* Include *separate* event hooks for redirects.
* Change the behaviour so that it fires the request/response event hooks, for all requests *including* redirects.

I *had* been planning on the first, but looking at it right now, today, I'd expect the second to *feel* more obvious, and it's certainly less complex to describe and is a simpler, smaller surface-area API.

*Either* way around we ought to document the behaviour.

It would also be a good idea for us to review what event_hooks behaviour `requests` has here.

**Reply by kummerer94** [reply]
Hi,

Thank you for the response! I really appreciate the time you invest into this.

I looked into your question around `requests`. I found that `requests` has [no `request` hook but only a `response` hook](https://docs.python-requests.org/en/master/user/advanced/#event-hooks).

Something similar to what you have implemented with the `httpx.HTTPTransport` or `httpx.AsyncHTTPTransport` is probably the [adapter logic in `requests`](https://docs.python-requests.org/en/master/user/advanced/#transport-adapters). These adapters (similar to the `HTTPTransport`) fire on both requests when the first one is a redirect.

What I finally did is look into the `response` hook that `requests` provides:

```python
import requests

def response_hook(r, *args, **kwargs):
    print(f"Status {r.status_code}")

requests.get("http://cogym.de", hooks={"response": response_hook})
```

The output of this is then:

```
Status 307
Status 200
<Response [200]>
```

So, the `response` hook is indeed called twice on redirect.

I would also prefer the `request` hook to fire twice when it encounters a redirect, simply because these are actually two requests. But I only recently started using your library and would not consider myself involved enough to have an informed opinion.

**Reply by lovelydinosaur** [reply]
Linking to the resolution... https://github.com/encode/httpx/pull/1806

Kinda a shame that discussions don't get backlinks when they're referenced in pull requests.

**Reply by kummerer94** [reply]
Already updated to the newest release version, tested it and it works great! Thanks again. It's a lot easier to work with hooks than with the transport classes due to obvious reasons...
